### ISSUE TITLE
Bug: [prefer-readonly] false positive with nested class accssing private members

### ISSUE THREAD

**Issue by timvahlbrock** [typescript-eslint/typescript-eslint#5230]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=4.7.2&sourceType=module&code=KYDwDg9gTgLgBAYwDYEMDOa4HkCuNhQDCqGcA3gLABQccYUAlgG4r5wDu6A6ozPgHZwAvHABmKJGmABuatVpgcAIyQMEcAObAYASX78CACgCU5ebUQR+aeBDwFhcGAAsGaWVXO0o2nFEHI6JiUNBYKyqrqACYQAMoQALbarvwaJmahYbR2+FAAdJxoPAx8wIIiMFA4Ml5ZCFZoEEjAeUgQaTkEBdy8AsYeWXAAvrUjoWNDQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Y6RAM0Wls4CGAEwD2TeIXRRe0EdEjgwAX2xgQSoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA

### Repro Code

```typescript
export class OuterClass {
  private wasWritten = false;

  public getInner() {
    const outer = this;

    return class {
      public doSomething() {
        outer.wasWritten = true;
        console.log(outer.wasWritten);
      }
    }
  }
}
```

### ESLint Config

```javascript
{
  "rules": {
    "@typescript-eslint/prefer-readonly": "error"
  }    
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

ESLint does not recommend to make `wasWritten` readonly, as it is written within the inner class.

### Actual Result

ESLint reports that `wasWritten` should be readonly. If this is done, the code no longer compiles.

### Additional Info

This issue does not occur when `wasWritten` is public.

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `5.29.0` |
| `@typescript-eslint/parser`            | `5.23.0` |
| `TypeScript`                           | `4.7.2` |
| `ESLint`                               | `8.15.0` |
| `node`                                 | `18.3.0` |

**Comment by bradzacher** [comment]
This is a very rare usecase to declare nested classes, let alone declare a nested class and mutate a parent class's property in the nested class.

Currently the rule assumes this doesn't happen - it assumes each class only touches its own properties. It would take some effort to fix this.
